File: packages/autojoin-bot/js/src/autojoin.js
import { cryptoBoxSeal } from "@serenity-kit/noble-sodium";

const CREDITS_PROGRAM = "credits.aleo";
const CREDITS_RECORD = "credits";
const USDCX_PROGRAM = "usdcx_stablecoin.aleo";
const USDCX_RECORD = "Token";

function arc20Family(name, program) {
  return Object.freeze({
    name,
    maxBatch: 15,
    recordPrograms: Object.freeze({
      mainnet: `${program}.aleo`,
      testnet: `test_${program}.aleo`,
    }),
    recordName: "Token",
    tokenIdentifiers: Object.freeze({ mainnet: program, testnet: `test_${program}` }),
    programs: Object.freeze({
      mainnet: Object.freeze(["main_aj_arc20_2_15.aleo"]),
      testnet: Object.freeze(["test_aj_arc20_2_15.aleo"]),
    }),
  });
}

export const JOIN_FAMILIES = Object.freeze({
  credits: Object.freeze({
    name: "credits",
    maxBatch: 16,
    recordPrograms: Object.freeze({ mainnet: CREDITS_PROGRAM, testnet: CREDITS_PROGRAM }),
    recordName: CREDITS_RECORD,
    programs: Object.freeze({
      mainnet: Object.freeze([
        "autojoin_credits_2_10.aleo",
        "autojoin_credits_11_14.aleo",
        "autojoin_credits_15_16.aleo",
      ]),
      testnet: Object.freeze([
        "autojoin_credits_2_10.aleo",
        "autojoin_credits_11_14.aleo",
        "autojoin_credits_15_16.aleo",
      ]),
    }),
  }),
  usdcx: Object.freeze({
    name: "usdcx",
    maxBatch: 16,
    recordPrograms: Object.freeze({
      mainnet: USDCX_PROGRAM,
      testnet: "test_usdcx_stablecoin.aleo",
    }),
    recordName: USDCX_RECORD,
    programs: Object.freeze({
      mainnet: Object.freeze([
        "aj_usdcx_stablecoin_2_10.aleo",
        "aj_usdcx_stablecoin_11_14.aleo",
        "aj_usdcx_stablecoin_15_16.aleo",
      ]),
      testnet: Object.freeze([
        "aj_test_usdcx_stablecoin_2_10.aleo",
        "aj_test_usdcx_stablecoin_11_14.aleo",
        "aj_test_usdcx_stablecoin_15_16.aleo",
      ]),
    }),
  }),
  arc20Eth: arc20Family("arc20-eth", "arc20_eth"),
  arc20Sol: arc20Family("arc20-sol", "arc20_sol"),
  arc20Wbtc: arc20Family("arc20-wbtc", "arc20_wbtc"),
});

export function joinCall(family, recordCount, network = "mainnet") {
  if (!Object.values(JOIN_FAMILIES).includes(family)) {
    throw new Error("unknown autojoin record family");
  }
  if (!Number.isInteger(recordCount) || recordCount < 2 || recordCount > family.maxBatch) {
    throw new Error(`${family.name} join size must be between 2 and ${family.maxBatch}`);
  }
  const programs = family.programs[network];
  if (!programs) throw new Error(`${family.name} autojoin is not configured for ${network}`);
  const programName = programs.length === 1
    ? programs[0]
    : recordCount <= 10
      ? programs[0]
      : recordCount <= 14
        ? programs[1]
        : programs[2];
  return { programName, functionName: `join_${recordCount}` };
}

export function creditsJoinCall(recordCount) {
  return joinCall(JOIN_FAMILIES.credits, recordCount);
}

export function usdcxJoinCall(recordCount, network = "mainnet") {
  return joinCall(JOIN_FAMILIES.usdcx, recordCount, network);
}

export function recordsForFamily(records, family, network = "mainnet") {
  const recordProgram = family.recordPrograms[network];
  if (!recordProgram) throw new Error(`${family.name} records are not configured for ${network}`);
  const selected = records.filter((record) => record.program_name === recordProgram
    && record.record_name === family.recordName);
  for (const record of selected) {
    if (typeof record.record_plaintext !== "string" || typeof record.tag !== "string") {
      throw new Error(`owned ${family.name} record is missing plaintext or tag`);
    }
  }
  return selected;
}

export function creditsRecords(records) {
  return recordsForFamily(records, JOIN_FAMILIES.credits);
}

export function canonicalProvingRequest(provingRequest) {
  const legacy = JSON.parse(provingRequest.toString());
  if (!legacy.authorization) throw new Error("SDK did not build an authorization proving request");
  return {
    broadcast: true,
    payload: {
      type: "authorization",
      authorization: legacy.authorization,
    },
  };
}

function acceptedBroadcast(result) {
  const value = result?.broadcast_result;
  return value?.status === "Accepted" || value?.status === "accepted"
    || (value && typeof value === "object" && "Accepted" in value);
}

export async function submitDelegated({ url, request, fetchImpl = fetch }) {
  const pubkeyResponse = await fetchImpl(`${url}/pubkey`);
  if (!pubkeyResponse.ok) {
    throw new Error(`Delegated-prover public-key request failed (HTTP ${pubkeyResponse.status}): ${await pubkeyResponse.text()}`);
  }
  const pubkey = await pubkeyResponse.json();
  const plaintext = new TextEncoder().encode(JSON.stringify(request));
  let ciphertext;
  try {
    ciphertext = cryptoBoxSeal({
      message: plaintext,
      publicKey: Uint8Array.from(Buffer.from(pubkey.public_key, "base64")),
    });
  } finally {
    plaintext.fill(0);
  }
  const cookie = pubkeyResponse.headers.get("set-cookie")?.split(";", 1)[0];
  const response = await fetchImpl(`${url}/prove`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...(cookie ? { Cookie: cookie } : {}),
    },
    body: JSON.stringify({
      key_id: pubkey.key_id,
      ciphertext: Buffer.from(ciphertext).toString("base64"),
    }),
  });
  const text = await response.text();
  if (!response.ok) {
    throw new Error(`Delegated proving failed (HTTP ${response.status}): ${text}`);
  }
  const result = JSON.parse(text);
  if (!acceptedBroadcast(result)) {
    throw new Error(`Delegated prover did not accept the broadcast: ${JSON.stringify(result.broadcast_result)}`);
  }
  return result;
}

const delay = (milliseconds) => new Promise((resolve) => setTimeout(resolve, milliseconds));

export async function consolidateRecords({
  family,
  network = "mainnet",
  sdk,
  privateKey,
  networkUrl,
  proverUrl,
  initialRecords,
  rescan,
  pollIntervalMs,
  timeoutMs,
  onScan = () => {},
  submit = submitDelegated,
}) {
  let records = initialRecords;
  let joins = 0;
  const manager = new sdk.ProgramManager(networkUrl);

  let dynamicProgramImports;
  while (recordsForFamily(records, family, network).length > 1) {
    const available = recordsForFamily(records, family, network);
    const count = Math.min(available.length, family.maxBatch);
    const selected = available.slice(0, count);
    const selectedTags = new Set(selected.map((record) => record.tag));
    const existingTags = new Set(available.map((record) => record.tag));
    const call = joinCall(family, count, network);
    const tokenIdentifier = family.tokenIdentifiers?.[network];
    if (tokenIdentifier && !dynamicProgramImports) {
      const dynamicProgram = family.recordPrograms[network];
      const dynamicProgramSource = await manager.networkClient.getProgram(dynamicProgram);
      dynamicProgramImports = await manager.networkClient.getProgramImports(dynamicProgramSource);
      dynamicProgramImports[dynamicProgram] = dynamicProgramSource;
    }
    const provingRequest = await manager.provingRequest({
      ...call,
      inputs: [
        ...(tokenIdentifier ? [`'${tokenIdentifier}'`] : []),
        ...selected.map((record) => record.record_plaintext),
      ],
      ...(dynamicProgramImports ? { programImports: dynamicProgramImports } : {}),
      privateKey,
      priorityFee: 0,
      privateFee: false,
      broadcast: true,
      useFeeMaster: true,
    });
    try {
      await submit({
        url: proverUrl,
        request: canonicalProvingRequest(provingRequest),
      });
    } finally {
      provingRequest.free?.();
    }
    joins += 1;

    const deadline = Date.now() + timeoutMs;
    while (true) {
      records = await rescan();
      onScan(records);
      const current = recordsForFamily(records, family, network);
      const inputsGone = current.every((record) => !selectedTags.has(record.tag));
      const replacementSeen = current.some((record) => !existingTags.has(record.tag));
      if (inputsGone && replacementSeen) break;
      if (Date.now() >= deadline) {
        throw new Error(`timed out waiting for ${call.programName}/${call.functionName} to reach the scanner`);
      }
      await delay(pollIntervalMs);
    }
  }
  return { records, joins, recordsRemaining: recordsForFamily(records, family, network).length };
}

export async function consolidateCredits(options) {
  const result = await consolidateRecords({ ...options, family: JOIN_FAMILIES.credits });
  return { ...result, creditsRemaining: result.recordsRemaining };
}

File: packages/autojoin-bot/rust/src/autojoin.rs
use std::{collections::HashMap, str::FromStr};

use anyhow::{Context, Result, anyhow, bail};
use base64::{Engine, engine::general_purpose::STANDARD};
use crypto_box::{PublicKey, aead::OsRng};
use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};
use snarkvm_circuit_network::Aleo;
use snarkvm_console::{
    account::PrivateKey,
    prelude::Network,
    program::{Identifier, Value as ProgramValue},
};
use snarkvm_synthesizer::{Process, Program};
use zeroize::Zeroizing;

use crate::{
    http::decode_response,
    network::AleoNetwork,
    records::{OwnedRecord, RecordFamily},
};

#[derive(Clone)]
pub struct DelegatedProverClient {
    client: Client,
    base_url: String,
}

#[derive(Debug, Deserialize)]
struct PubkeyResponse {
    key_id: String,
    public_key: String,
}

#[derive(Serialize)]
struct AuthorizationProvingRequest<T> {
    broadcast: bool,
    payload: AuthorizationPayload<T>,
}

#[derive(Serialize)]
struct AuthorizationPayload<T> {
    #[serde(rename = "type")]
    payload_type: &'static str,
    authorization: T,
}

fn authorization_proving_request<T>(authorization: T) -> AuthorizationProvingRequest<T> {
    AuthorizationProvingRequest {
        broadcast: true,
        payload: AuthorizationPayload {
            payload_type: "authorization",
            authorization,
        },
    }
}

impl DelegatedProverClient {
    pub fn new(base_url: String) -> Self {
        Self {
            client: Client::new(),
            base_url,
        }
    }

    fn pubkey_url(&self) -> String {
        format!("{}/pubkey", self.base_url)
    }

    fn prove_url(&self) -> String {
        format!("{}/prove", self.base_url)
    }

    pub async fn prove_and_broadcast<N: Network, A: Aleo<Network = N>>(
        &self,
        private_key: &PrivateKey<N>,
        network: AleoNetwork,
        family: RecordFamily,
        records: &[&OwnedRecord],
    ) -> Result<Value> {
        let (program_id, function_name) =
            join_call(family, network, records.len()).with_context(|| {
                format!(
                    "{} join size must be between 2 and {} for {}",
                    family.name(),
                    family.max_batch(),
                    network.as_str()
                )
            })?;
        let mut roots = vec![program_id];
        if family.token_identifier(network).is_some() {
            roots.push(
                family
                    .record_program(network)
                    .context("ARC20 autojoin is not configured for this network")?,
            );
        }
        let programs = self.fetch_program_graph::<N>(network, &roots).await?;
        let authorization = authorize::<N, A>(
            private_key,
            family,
            network,
            records,
            &programs,
            program_id,
            function_name,
        )?;
        let request = authorization_proving_request(authorization);
        self.submit(request).await
    }

    async fn fetch_program_graph<N: Network>(
        &self,
        network: AleoNetwork,
        root_programs: &[&str],
    ) -> Result<Vec<Program<N>>> {
        let mut pending = root_programs
            .iter()
            .map(|program| (*program).to_owned())
            .collect::<Vec<_>>();
        let mut programs = HashMap::new();

        while let Some(program_id) = pending.pop() {
            if program_id == "credits.aleo" || programs.contains_key(&program_id) {
                continue;
            }
            let source_url = format!(
                "https://api.explorer.provable.com/v2/{}/program/{program_id}",
                network.as_str()
            );
            let response = self.client.get(source_url).send().await?;
            let source: String = decode_response(response, "Autojoin-program fetch").await?;
            let program = Program::<N>::from_str(&source)
                .with_context(|| format!("invalid deployed program source for {program_id}"))?;
            pending.extend(program.imports().keys().map(ToString::to_string));
            programs.insert(program_id, program);
        }

        topological_programs(programs)
    }

    async fn submit(&self, request: impl Serialize) -> Result<Value> {
        let pubkey_response = self.client.get(self.pubkey_url()).send().await?;
        let cookie = pubkey_response
            .headers()
            .get(reqwest::header::SET_COOKIE)
            .and_then(|value| value.to_str().ok())
            .and_then(|value| value.split(';').next())
            .map(str::to_owned);
        let pubkey: PubkeyResponse =
            decode_response(pubkey_response, "Delegated-prover public-key request").await?;
        let public_key: [u8; 32] = STANDARD
            .decode(pubkey.public_key)?
            .try_into()
            .map_err(|_| anyhow!("delegated prover public key must be 32 bytes"))?;
        let plaintext = Zeroizing::new(serde_json::to_vec(&request)?);
        let ciphertext = PublicKey::from(public_key)
            .seal(&mut OsRng, &plaintext)
            .map_err(|error| anyhow!("failed to encrypt delegated proving request: {error}"))?;

        let prove_request = self.client.post(self.prove_url()).json(&json!({
            "key_id": pubkey.key_id,
            "ciphertext": STANDARD.encode(ciphertext),
        }));
        let prove_request = if let Some(cookie) = cookie {
            prove_request.header(reqwest::header::COOKIE, cookie)
        } else {
            prove_request
        };
        let result: Value =
            decode_response(prove_request.send().await?, "Delegated proving").await?;
        if !broadcast_accepted(&result) {
            bail!(
                "delegated prover did not accept the broadcast: {}",
                result["broadcast_result"]
            );
        }
        Ok(result)
    }
}

fn authorize<N: Network, A: Aleo<Network = N>>(
    private_key: &PrivateKey<N>,
    family: RecordFamily,
    network: AleoNetwork,
    records: &[&OwnedRecord],
    programs: &[Program<N>],
    root_program: &str,
    function_name: &str,
) -> Result<snarkvm_synthesizer::Authorization<N>> {
    let program = programs
        .iter()
        .find(|program| program.id().to_string() == root_program)
        .context("autojoin root program is missing")?;
    let process = Process::<N>::load().context("failed to initialize authorization process")?;
    for dependency in programs {
        process
            .lock()
            .add_program(dependency)
            .with_context(|| format!("failed to load {}", dependency.id()))?;
    }
    let mut inputs =
        Vec::with_capacity(records.len() + usize::from(family.token_identifier(network).is_some()));
    if let Some(token_identifier) = family.token_identifier(network) {
        inputs.push(
            ProgramValue::<N>::from_str(&format!("'{token_identifier}'"))
                .context("invalid ARC20 token program identifier")?,
        );
    }
    inputs.extend(
        records
            .iter()
            .map(|record| {
                let plaintext = record
                    .record_plaintext
                    .as_deref()
                    .context("owned record is missing a decrypted plaintext")?;
                ProgramValue::<N>::from_str(plaintext)
                    .context("scanner returned an invalid record plaintext")
            })
            .collect::<Result<Vec<_>>>()?,
    );
    process
        .authorize::<A, _>(
            private_key,
            program.id(),
            Identifier::<N>::from_str(function_name)?,
            inputs.iter(),
            &mut rand::rng(),
        )
        .context("failed to authorize autojoin call")
}

fn topological_programs<N: Network>(
    mut programs: HashMap<String, Program<N>>,
) -> Result<Vec<Program<N>>> {
    let mut ordered = Vec::with_capacity(programs.len());
    let mut loaded = std::collections::HashSet::from(["credits.aleo".to_owned()]);

    while !programs.is_empty() {
        let ready = programs.iter().find_map(|(id, program)| {
            program
                .imports()
                .keys()
                .all(|dependency| loaded.contains(&dependency.to_string()))
                .then(|| id.clone())
        });
        let id = ready.context("autojoin program imports contain a cycle or missing dependency")?;
        let program = programs.remove(&id).expect("selected from map");
        loaded.insert(id);
        ordered.push(program);
    }

    Ok(ordered)
}

fn broadcast_accepted(result: &Value) -> bool {
    result.get("broadcast_result").is_some_and(|broadcast| {
        broadcast.get("Accepted").is_some()
            || broadcast
                .get("status")
                .and_then(Value::as_str)
                .is_some_and(|status| status.eq_ignore_ascii_case("accepted"))
    })
}

pub const fn credits_join_call(count: usize) -> Option<(&'static str, &'static str)> {
    join_call(RecordFamily::Credits, AleoNetwork::Mainnet, count)
}

pub const fn join_call(
    family: RecordFamily,
    network: AleoNetwork,
    count: usize,
) -> Option<(&'static str, &'static str)> {
    let program = match (family, network, count) {
        (RecordFamily::Credits, _, 2..=10) => "autojoin_credits_2_10.aleo",
        (RecordFamily::Credits, _, 11..=14) => "autojoin_credits_11_14.aleo",
        (RecordFamily::Credits, _, 15..=16) => "autojoin_credits_15_16.aleo",
        (RecordFamily::Usdcx, AleoNetwork::Mainnet, 2..=10) => "aj_usdcx_stablecoin_2_10.aleo",
        (RecordFamily::Usdcx, AleoNetwork::Mainnet, 11..=14) => "aj_usdcx_stablecoin_11_14.aleo",
        (RecordFamily::Usdcx, AleoNetwork::Mainnet, 15..=16) => "aj_usdcx_stablecoin_15_16.aleo",
        (RecordFamily::Usdcx, AleoNetwork::Testnet, 2..=10) => "aj_test_usdcx_stablecoin_2_10.aleo",
        (RecordFamily::Usdcx, AleoNetwork::Testnet, 11..=14) => {
            "aj_test_usdcx_stablecoin_11_14.aleo"
        }
        (RecordFamily::Usdcx, AleoNetwork::Testnet, 15..=16) => {
            "aj_test_usdcx_stablecoin_15_16.aleo"
        }
        (
            RecordFamily::Arc20Eth | RecordFamily::Arc20Sol | RecordFamily::Arc20Wbtc,
            AleoNetwork::Mainnet,
            2..=15,
        ) => "main_aj_arc20_2_15.aleo",
        (
            RecordFamily::Arc20Eth | RecordFamily::Arc20Sol | RecordFamily::Arc20Wbtc,
            AleoNetwork::Testnet,
            2..=15,
        ) => "test_aj_arc20_2_15.aleo",
        _ => return None,
    };
    let function = match count {
        2 => "join_2",
        3 => "join_3",
        4 => "join_4",
        5 => "join_5",
        6 => "join_6",
        7 => "join_7",
        8 => "join_8",
        9 => "join_9",
        10 => "join_10",
        11 => "join_11",
        12 => "join_12",
        13 => "join_13",
        14 => "join_14",
        15 => "join_15",
        16 => "join_16",
        _ => return None,
    };
    Some((program, function))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn edge_routes_are_network_scoped() {
        let client = DelegatedProverClient::new(AleoNetwork::Mainnet.prover_endpoint());
        assert_eq!(
            client.pubkey_url(),
            "https://edge.provable.com/api/prove/mainnet/pubkey"
        );
        assert_eq!(
            client.prove_url(),
            "https://edge.provable.com/api/prove/mainnet/prove"
        );
    }

    #[test]
    fn proving_request_omits_optional_job_id_for_edge_compatibility() {
        let request = serde_json::to_value(authorization_proving_request(json!({
            "requests": [],
            "transitions": [],
        })))
        .unwrap();

        assert_eq!(request["broadcast"], true);
        assert_eq!(request["payload"]["type"], "authorization");
        assert!(request.get("job_id").is_none());
    }

    use snarkvm_console::prelude::TestnetV0;

    #[test]
    fn join_program_bands_are_exact() {
        assert_eq!(
            credits_join_call(2),
            Some(("autojoin_credits_2_10.aleo", "join_2"))
        );
        assert_eq!(
            credits_join_call(10).unwrap().0,
            "autojoin_credits_2_10.aleo"
        );
        assert_eq!(
            credits_join_call(11).unwrap().0,
            "autojoin_credits_11_14.aleo"
        );
        assert_eq!(
            credits_join_call(14).unwrap().0,
            "autojoin_credits_11_14.aleo"
        );
        assert_eq!(
            credits_join_call(15).unwrap().0,
            "autojoin_credits_15_16.aleo"
        );
        assert_eq!(
            credits_join_call(16),
            Some(("autojoin_credits_15_16.aleo", "join_16"))
        );
        assert_eq!(credits_join_call(1), None);
        assert_eq!(credits_join_call(17), None);
        assert_eq!(
            join_call(RecordFamily::Usdcx, AleoNetwork::Mainnet, 2),
            Some(("aj_usdcx_stablecoin_2_10.aleo", "join_2"))
        );
        assert_eq!(
            join_call(RecordFamily::Usdcx, AleoNetwork::Testnet, 11)
                .unwrap()
                .0,
            "aj_test_usdcx_stablecoin_11_14.aleo"
        );
        assert_eq!(
            join_call(RecordFamily::Usdcx, AleoNetwork::Testnet, 16),
            Some(("aj_test_usdcx_stablecoin_15_16.aleo", "join_16"))
        );
        assert_eq!(
            join_call(RecordFamily::Arc20Eth, AleoNetwork::Mainnet, 15),
            Some(("main_aj_arc20_2_15.aleo", "join_15"))
        );
        assert_eq!(
            join_call(RecordFamily::Arc20Wbtc, AleoNetwork::Mainnet, 16),
            None
        );
        assert_eq!(
            join_call(RecordFamily::Arc20Sol, AleoNetwork::Testnet, 2),
            Some(("test_aj_arc20_2_15.aleo", "join_2"))
        );
    }

    #[test]
    fn import_graph_is_ordered_before_root() {
        let dependency = Program::<TestnetV0>::from_str(
            "program dependency.aleo;\n\nfunction noop:\n    input r0 as u8.public;\n    output r0 as u8.public;\n",
        )
        .unwrap();
        let root = Program::<TestnetV0>::from_str(
            "import dependency.aleo;\n\nprogram root.aleo;\n\nfunction noop:\n    input r0 as u8.public;\n    output r0 as u8.public;\n",
        )
        .unwrap();
        let programs = HashMap::from([
            ("root.aleo".to_owned(), root),
            ("dependency.aleo".to_owned(), dependency),
        ]);
        let ordered = topological_programs(programs).unwrap();
        assert_eq!(ordered[0].id().to_string(), "dependency.aleo");
        assert_eq!(ordered[1].id().to_string(), "root.aleo");
    }

    #[test]
    fn arc20_program_identifier_is_a_valid_public_input() {
        assert!(ProgramValue::<TestnetV0>::from_str("'arc20_eth'").is_ok());
    }
}
